src/production/phases/CodeSplittingPhase.ts

```diff
@@ -0,0 +1,189 @@
+import { Module } from '../../core/Module';
+import { IProductionContext } from '../ProductionContext';
+import { ImportType } from '../module/ImportReference';
+import { SplitEntry, ISplitEntry } from '../module/SplitEntries';
+
+/**
+ * Function to check if a module is only required thr
+ * @param possibleSplitEntry
+ */
+function resolveDynamicImport(possibleSplitEntry: Module): boolean {
+  const { moduleTree } = possibleSplitEntry;
+  let isDynamic = true;
+  for (const dependant of moduleTree.dependants) {
+    // if all dependants require this module dynamic
+    // then we have a splitEntry
+    if (dependant.type !== ImportType.DYNAMIC_IMPORT) {
+      isDynamic = false;
+      break;
+    }
+  }
+  return isDynamic;
+}
+
+type ModuleCache = {
+  [key: number]: boolean;
+};
+
+/**
+ * The core of the code splitting
+ *
+ * Provide it with a target and it will spit out a valid ISplitEntry
+ *
+ * @param productionContext
+ * @param target
+ */
+export function ResolveSplitEntry(productionContext: IProductionContext, target: Module): ISplitEntry {
+  const entryModuleId = target.moduleId;
+  const subModules: Array<Module> = [target];
+  const visited: ModuleCache = {};
+  const traversed: ModuleCache = {
+    [entryModuleId]: true,
+  };
+
+  const splitEntry = {
+    entry: target,
+    resolve: function (): ISplitEntry {
+      this.traverseDependencies(target, true);
+
+      return SplitEntry({
+        module: this.entry,
+        productionContext,
+        subModules: this.subModules,
+      });
+    },
+    subModules,
+    /**
+     * traceCircularDependency accepts a target of type Module
+     * it needs a reference to the parentId so we can safely trace back
+     *
+     * @param target
+     * @param parentId
+     */
+    traceCircularDependency: function (target: Module, parentId: number): boolean {
```

Comment:
I can merge this function into traceOrigin...

Downside, traceOrgin will look more complicated and harder to understand in the future...

@nchanged any preference?

Reply:
I like the way it is now ;-)

Reply:
✅ me too